synapse/lib/stormlib/macro.py

```diff
@@ -0,0 +1,184 @@
+            ('storm', {'help': 'The storm command string or embedded query to set.'}),
+        ),
+        'storm': '''
+            $lib.macro.set($cmdopts.name, $cmdopts.storm)
+            $lib.print('Set macro: {name}', name=$cmdopts.name)
+        ''',
+    },
+    {
+        'name': 'macro.get',
+        'descr': macro_get_descr,
+        'cmdargs': (
+            ('name', {'help': 'The name of the macro to display.'}),
+        ),
+        'storm': '''
+            $mdef = $lib.macro.get($cmdopts.name)
+            if $mdef {
+                $lib.print($mdef.storm)
+            } else {
+                $lib.print('Macro not found: {name}', name=$cmdopts.name)
+            }
+        ''',
+    },
+    {
+        'name': 'macro.list',
+        'descr': macro_list_descr,
+        'storm': '''
+            $count = $(0)
+            for ($name, $mdef) in $lib.macro.list() {
+                $user = $lib.auth.users.get($mdef.user)
+                $lib.print('{name} (owner: {user})', name=$name.ljust(20), user=$user.name)
+                $count = $($count + 1)
+            }
+            $lib.print('{count} macros found', count=$count)
+        ''',
+    },
+]
+
+class MacroExecCmd(s_storm.Cmd):
+    # FIXME Give a description of use.
+    '''
+    Execute a named macro.
+    '''
+
+    name = 'macro.exec'
+
+    def getArgParser(self):
+        pars = s_storm.Cmd.getArgParser(self)
+        pars.add_argument('name', help='The name of the macro to execute')
+        return pars
+
+    async def execStormCmd(self, runt, genr):
+
+        if not self.runtsafe:
+            mesg = 'macro.exec does not support per-node invocation'
+            raise s_exc.StormRuntimeError(mesg=mesg)
+
+        name = await s_stormtypes.tostr(self.opts.name)
+        path = ('cortex', 'storm', 'macros', name)
+
+        mdef = await runt.snap.core.getHiveKey(path)
+        if mdef is None:
+            # FIXME Bad ref "name"
+            mesg = f'Macro name not found: {name}'
+            raise s_exc.NoSuchName(mesg=mesg)
+
+        query = await runt.getStormQuery(mdef['storm'])
+
+        with runt.snap.getStormRuntime() as subr:
+
+            async def wrapgenr():
+
+                async for node, path in genr:
+                    path.initframe(initrunt=subr)
+                    yield node, path
+
+            subr.loadRuntVars(query)
+
+            async for node, path in subr.iterStormQuery(query, genr=wrapgenr()):
+                path.finiframe(runt)
+                yield node, path
+
+class LibMacro(s_stormtypes.Lib):
+
+    def addLibFuncs(self):
+        self.locls.update({
+            'set': self._funcMacroSet,
+            'get': self._funcMacroGet,
+            'del': self._funcMacroDel,
+            'list': self._funcMacroList,
+        })
+
+    async def _funcMacroList(self):
+        path = ('cortex', 'storm', 'macros')
+        return await self.runt.snap.core.getHiveKeys(path)
```

Comment:
Attempting to use this with with spawn ends up with an attribute error:
`ERROR: ('AttributeError', {'efile': 'macro.py', 'eline': 136, 'mesg': "'SpawnCore' object has no attribute 'getHiveKeys'"})`